common/dictionary-db/dictionary-db.ts

```diff
@@ -0,0 +1,1416 @@
+                    await this._buildIdHealthCheck(buildId, trackBuildIdsToClear);
+                    await this._saveTokensForDB(
+                        profile,
+                        trackStates,
+                        records,
+                        ankiCards,
+                        modifiedCardsBatch,
+                        trackTokensMap,
+                        modifiedTokens
+                    );
+                    await this._gatherModifiedTokens(profile, modifiedTokens);
+                });
+
+                progress.current += modifiedCardsBatch.size;
+                await this._updateBuildAnkiCacheProgress(
+                    buildId,
+                    trackBuildIdsToClear,
+                    progress,
+                    Array.from(modifiedTokens),
+                    statusUpdates,
+                    t
+                );
+            },
+            { batchSize: 100 } // Batch for memory usage and yomitan cache size
+        );
+    }
+
+    private async _saveTokensForDB(
+        profile: string,
+        trackStates: Map<number, TrackStateForDB>,
+        records: DictionaryTokenRecord[],
+        ankiCards: DictionaryAnkiCardRecord[],
+        modifiedCardsBatch: CardsForDB,
+        trackTokensMap: Map<
+            number,
+            Map<DictionaryTokenSource, Map<string, { lemmas: string[]; cardIds: Set<number> }>>
+        >,
+        modifiedTokens: Set<string>
+    ): Promise<void> {
+        for (const record of records) {
+            modifiedTokens.add(record.token);
+            for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+        }
+        return this.db.transaction('rw', this.db.tokens, this.db.ankiCards, async () => {
+            await Promise.all([this._saveRecordBulk(records), this.db.ankiCards.bulkPut(ankiCards)]);
+            await this.db.tokens
+                .where('cardIds')
+                .anyOf(Array.from(modifiedCardsBatch.keys()))
+                .distinct()
+                .filter((r) => r.profile === profile)
+                .modify((record, ref) => {
+                    if (
+                        record.source !== DictionaryTokenSource.ANKI_WORD &&
+                        record.source !== DictionaryTokenSource.ANKI_SENTENCE
+                    ) {
+                        return;
+                    }
+                    if (!trackStates.has(record.track)) return;
+                    if (trackTokensMap.get(record.track)!.get(record.source)!.has(record.token)) return; // We want tokens that were not updated but refers to updated cards (e.g. field value changed, different tokens)
+                    const validCardIds = record.cardIds.filter((id) => !modifiedCardsBatch.has(id));
+                    if (!validCardIds.length) {
+                        modifiedTokens.add(record.token);
+                        for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+                        delete (ref as any).value;
```

Comment:
I commented on a similar block of code somewhere else but if it's easy to do then I think you'll squeeze more readability out of the code by reducing duplication. For example, here the `validCardIds` checks could be used to branch on the modification to the record, without duplicating the updates to `modifiedTokens`.